tools/codegen/api/cpp.py

```diff
@@ -102,7 +102,7 @@ def argumenttype_type(t: Type, *, mutable: bool, binds: ArgName, remove_non_owni
             else:
                 return NamedCType(binds, BaseCType(intArrayRefT))
         elif str(t.elem) == 'Tensor':
-            return NamedCType(binds, BaseCType(tensorListT))
+            return NamedCType(binds, BaseCType(iTensorListT))
```

Comment:
Hmm, so I'm not sure that we want this. Right now, the unboxed C++ API for an op like `torch.cat` looks like this:
```
TORCH_API inline at::Tensor cat(at::TensorList tensors, int64_t dim=0);
```

And this change would turn it into this:
```
TORCH_API inline at::Tensor cat(at::ITensorList tensors, int64_t dim=0)
```

That might be fine...? I think it's a little hairy because we'd technically like to support passing a bunch of variations into the C++ API:
- `ArrayRef<Tensor>` (this is the same thing as `TensorList`, aka the current state)
- initializer list: `{t1, t2}` (this is convertible into an `ArrayRef<Tensor>`)
- `c10::List<Tensor>` (this would be a new API)

I think it might be clearer just to generate a new overload explicitly for `c10::List`.

Reply:
> I think it's a little hairy because we'd technically like to support passing a bunch of variations into the C++ API, which would mean supporting all of the implicit constructors without any ambiguities

I think I'm already doing that, though. Am I missing something here?

> I think it might be clearer just to generate a new overload explicitly for `c10::List`

That also sounds good. So, here's what I was planning to do:

1. This PR introduces `ITensorList`, and enable structured kernel support
2. The next one ports `cat` to structured kernels
3. Then, in the PR after that, use it in the C++ API
4. Finally, work on supporting `Tensor[]?`, too

What do you think?

Reply:
@bdhirsh
> I think it might be clearer just to generate a new overload explicitly for `c10::List`

I was thinking about this, and do you still think we should generate a new overload? AFAIU, `ITensorList` already covers all constructor cases of `ArrayRef`. Am I missing anything?

Reply:
So you're proposing that the C++ API should look like this, right?
```
TORCH_API inline at::Tensor cat(at::ITensorList tensors, int64_t dim=0)
```

And that if I try to call `cat(TensorList)`, or `cat(c10::List<Tensor>)`, in both cases we'll call one of `ITensorList`'s implicit constructors.

Sure! I guess the only downside I can think of is that we won't be able to add implicit constructors from `ITensorList` -> `TensorList` or `List<Tensor>`. But I'm not sure that we'd want that anyway.

Reply:
> Sure! I guess the only downside I can think of is that we won't be able to add implicit constructors from ITensorList -> TensorList or List<Tensor>. But I'm not sure that we'd want that anyway.

You are right! That may be one thing to work on, since changing `cpp.py` would mean that `Tensor[]` output would get converted to `ITensorList` (not sure if that's what happens), right?

Anyways, I think, for this PR, I will stick to make structured kernels work with tensor lists. How does that sound?

Reply:
> since changing cpp.py would mean that Tensor[] output would get converted to ITensorList (not sure if that's what happens), right?

Oh yeah, thanks for reminding me :). That's probably the better reason to not make the change in `cpp.py`. In theory we could deal with that by having that translation function check if the current native_function is structured, but that would really complicate that logic unnecessarily.

>  I will stick to make structured kernels work with tensor lists. How does that sound?

Sounds good to me! So The C++ API knows nothing of `ITensorList`, and it only shows up in the structured kernel schemas.